### PR TITLE
Reusable blocks: Improve UX for non-privileged users

### LINKED ISSUE #12338: Reusable blocks: Bad UX for non-privileged users
There are some very rough edges when using, creating and modifying reusable blocks as an author or contributor. I've summarised the issues here as it's easier to address them as one task.

#### Creating a reusable block as a contributor

Expected: The _Add to Reusable Blocks_ button should be disabled.
Actual: The _Add to Reusable Blocks_ button is enabled. Clicking it results in an error notice appearing when the API request fails.

#### Updating another user's reusable block as an author

Expected: The _Edit_ button should be disabled.
Actual: The _Edit_ button is enabled. Clicking it lets you edit the block and save changes which results in an error notice appearing when the API request fails.

#### Deleting another user's reusable block as an author or contributor

Expected: The _Remove from Reusable Blocks_ button should be disabled.
Actual: The  _Remove from Reusable Blocks_ button is enabled. Clicking it results in an error notice appearing when the API request fails.

### PR DESCRIPTION
**Pull request by noisysocks**
### What this is

Fixes #12338.

Improves the UX of creating, editing, and deleting a reusable block when logged in as an author or contributor by disabling the _Add to Reusable Blocks_, _Edit_, and _Remove from Reusable Blocks_ buttons when necessary.

This is accomplished under the hood by introducing the `canUser()` selector to `core-data` which allows callers to query whether the REST API supports performing a given action on a given resource, e.g. one can query whether the logged in user can create posts by running `wp.data.select( 'core' ).canUser( 'create', 'posts' )`.

The existing `hasUploadPermissions()` selector is changed to use `canUser( 'create', 'media' )` under the hood.

### How it looks

_Add to Reusable Blocks_ button is visible for admins and editors:

admin - add

_Add to Reusable Blocks_ button is hidden for contributors:

contributor - add

_Edit_ button is enabled and _Remove from Reusable Blocks_ is visible for admins and editors:

admin - edit remove

_Edit_ button is disabled and _Remove from Reusable Blocks_ is hidden for authors:

author - edit remove

### How to test

1. Try to create a reusable block as a contributor. The _Add to Reusable Blocks_ button should be disabled.
1. Try to updating another user's reusable block as an author. The _Edit_ button should be disabled.
1. Try to deleting another user's reusable block as an author or contributor. The _Remove from Reusable Blocks_ button should be disabled.

### CODE REVIEW

**Comment by noisysocks** [comment]
This should be ready!

@mtias: Reckon we could plop this back into 4.7? I'd like to have this fixed for 5.0.1 since it's a pretty awful UX otherwise. Not sure where we are timing wise.

File: packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js
Comment (talldan) [inline packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js]: I think this is the only block settings menu item that ends up disabled. Most of the others are just not rendered when they're not valid in a specific context and I think it's worth following the same pattern here. The same pattern (hiding instead of disabling) is also used for other bits of UI when users don't have the capability.

File: packages/core-data/src/selectors.js
Comment (talldan) [inline packages/core-data/src/selectors.js]: I like the expressiveness of the canUser selector 😄

File: packages/editor/src/components/block-settings-menu/reusable-block-delete-button.js
Comment (talldan) [inline packages/editor/src/components/block-settings-menu/reusable-block-delete-button.js]: I think same comment here regarding disabling vs. hiding.

Comment (talldan) [reply packages/editor/src/components/block-settings-menu/reusable-block-delete-button.js]: Also noting that in testing I didn't see the flash of this switching from enabled to disabled. That seems to be because the GET request for the block has already been made and cached. If an OPTIONS request was made instead of a GET request the issue would resurface.

File: packages/core-data/src/resolvers.js
Comment (talldan) [inline packages/core-data/src/resolvers.js]: Is that just for /post/:id of for the all resources? I mentioned that because the logic below happens for all resources.

Is there a ticket/issue covering the bug?

**Review by talldan — DISMISSED** [review]
I like the canUser function, it's a nice way to express that functionality.

I think that flash of the disabled state of the block settings menu item is something that needs to be fixed. 

Preloading could work, but I think checking the resolution state is also possible and worth doing as well, since it shouldn't just depend on preloading. I think this is what embeds do:
https://github.com/WordPress/gutenberg/blob/master/packages/core-data/src/selectors.js#L41

File: packages/core-data/src/selectors.js
Comment (noisysocks) [reply packages/core-data/src/selectors.js]: Props to @gziolo for suggesting the API 🙂

File: packages/core-data/src/resolvers.js
Comment (noisysocks) [reply packages/core-data/src/resolvers.js]: I've only checked posts but I suspect the bug affects all REST endpoints. Will do some investigating and open up a Trac ticket 👍

File: packages/core-data/src/resolvers.js
Comment (noisysocks) [inline packages/core-data/src/resolvers.js]: Should I mark this as deprecated? I'm unsure on what the policy is post 5.0 🙂

File: packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js
Comment (noisysocks) [reply packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js]: 👌 I've added this OPTIONS request to the list of requests we preload in 0115ae6a7.

Noting that this is a change we'll need to make to core after this PR is merged.

File: packages/core-data/src/selectors.js
Comment (noisysocks) [reply packages/core-data/src/selectors.js]: 👌 537661450ac04c7d4867c424887dd64fd225b672

File: packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js
Comment (noisysocks) [reply packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js]: 👌 I've added this OPTIONS request to the list of requests we preload in bc0709f27cfa4fef3b30847c7aa328a69850799f.

Noting that this is a change we'll need to make to core after this PR is merged.

File: packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js
Comment (noisysocks) [reply packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js]: > The same pattern (hiding instead of disabling) is also used for other bits of UI when users don't have the capability.

Makes sense to be consistent here then 🙂 — have updated the behaviour in 708a50484fc99537a5f3368b4dbb66e134198741.

File: packages/editor/src/components/block-settings-menu/reusable-block-delete-button.js
Comment (noisysocks) [reply packages/editor/src/components/block-settings-menu/reusable-block-delete-button.js]: Have changed this to remove the button instead of disabling in 708a50484fc99537a5f3368b4dbb66e134198741.

We can't preload these requests because we don't know the reusable block ID ahead of time.

File: packages/core-data/src/resolvers.js
Comment (talldan) [reply packages/core-data/src/resolvers.js]: I don't know either. Might be worth pinging someone on slack. Will hold off approving until the answer is known.

File: packages/editor/src/components/block-settings-menu/reusable-block-delete-button.js
Comment (talldan) [reply packages/editor/src/components/block-settings-menu/reusable-block-delete-button.js]: Yep, I do understand that we can't preload. My worry is that it's only by coincidence that the user's capability is known ahead of time. If the underlying code switches to use OPTIONS instead of GET, users would start seeing this button initially be clickable but then suddenly disappear.

This is probably a nitpick though and not a blocker. I'm not even sure what the best UX would be.

File: packages/core-data/src/resolvers.js
Comment (noisysocks) [reply packages/core-data/src/resolvers.js]: cc. @youknowriad

Comment (noisysocks) [reply packages/core-data/src/resolvers.js]: It's probably fine leaving this as is where we support both `canUser` and `hasUploadPermissions`. We can always deprecate later.

**Comment by noisysocks** [comment]
@gziolo any thoughts on this one?

File: packages/editor/src/components/block-settings-menu/reusable-block-delete-button.js
Comment (noisysocks) [reply packages/editor/src/components/block-settings-menu/reusable-block-delete-button.js]: Ah, got it! I wonder if defaulting to disabled would be a better experience...

I think that, further down the track, `core-data` should modify the `canUser` state whenever it receives an `Allow` header from _any_ response, and not just when the `canUser()` resolver is called.

Comment (talldan) [reply packages/editor/src/components/block-settings-menu/reusable-block-delete-button.js]: Yeah, that's a nice idea. Not sure how it'd work, but a nice idea 😄

File: packages/core-data/src/resolvers.js
Comment (youknowriad) [reply packages/core-data/src/resolvers.js]: Yes, for now, we should leave those, we can add a deprecation message if needed but without specifying a version. 

This is something we should discuss more in the next #core-js chats.

File: packages/editor/src/components/block-settings-menu/reusable-block-delete-button.js
Comment (gziolo) [inline packages/editor/src/components/block-settings-menu/reusable-block-delete-button.js]: `select` - that's interesting to see it have much broader application than I anticipated. I hope there are no drawbacks of that approach :)

File: packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js
Comment (gziolo) [inline packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js]: It's probably better to move `isVisible` outside of the object - the way it was before. It makes return value harder follow at the moment.

File: packages/core-data/src/resolvers.js
Comment (gziolo) [reply packages/core-data/src/resolvers.js]: If this is the case, it should be fixed /cc @danielbachhuber @kadamwhite

File: packages/core-data/src/selectors.js
Comment (gziolo) [inline packages/core-data/src/selectors.js]: Can we add the information about the default behavior when it is not found in the store?

**Comment by gziolo** [comment]
> The existing `hasUploadPermissions()` selector is changed to use `canUser( 'create', 'media' )` under the hood.

@imath - I think you implemented the original version of this behavior when working on media upload permissions. Can you double check if the generalized version is still correct?

@danielbachhuber or @kadamwhite - it would be nice to get your review on the bits that interact with REST API.

File: packages/core-data/src/resolvers.js
Comment (TimothyBJacobs) [reply packages/core-data/src/resolvers.js]: If you just want the `Allow` header, what about doing a `HEAD` request instead of `GET`.

Comment (danielbachhuber) [reply packages/core-data/src/resolvers.js]: Looking into this further:

1. The `Allow` header is generated by `rest_send_allow_header()`, which is called on `rest_post_dispatch`.
    * https://github.com/WordPress/WordPress/blob/d2782aabc51f387c405eacc2ddbd38dfdacd4405/wp-includes/rest-api.php#L628
    * https://github.com/WordPress/WordPress/blob/d2782aabc51f387c405eacc2ddbd38dfdacd4405/wp-includes/rest-api.php#L167
2. `rest_post_dispatch` is called for every response (I think):
    * https://github.com/WordPress/WordPress/blob/47d32decd6668b3ccf100acc8a40aa0a9fc35003/wp-includes/rest-api/class-wp-rest-server.php#L352

It seems like `Allow` should be present for `OPTIONS`. It's an officially supported header: https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/OPTIONS

Can you provide steps to reproduce?

**Comment by danielbachhuber** [comment]
> Improves the UX of creating, editing, and deleting a reusable block when logged in as an author or contributor by disabling the _Add to Reusable Blocks_, _Edit_, and _Remove from Reusable Blocks_ buttons when necessary.

Does this change the UI at all? If so, can you include screenshots and/or GIFs in the description?

**Comment by danielbachhuber** [comment]
Also:
> The only important thing from my perspective is to ensure that `canUser` has the best strategy for returning the default value.

At first glance, I'm not sure how well this `canUser` abstraction is going to work in practice. WordPress' roles and capabilities system is quite nuanced, and abstractions on top of it can be problematic. Just want to flag that.

**Comment by imath** [comment]
@gziolo Hi,

I've just tested the PR. I confirm logged in as an Administrator I'm still able to upload files and that being logged in as a contributor I can only use external URLs to add media. So it seems to be working as expected for this part 👍 

@noisysocks FYI I also tested the Reusable blocks UI changes being logged in as a contributor: 
- I confirm I cannot create new reusable block.
- I can use another user reusable block but i cannot edit it.
- I can access to the `/wp-admin/edit.php?post_type=wp_block` manage screen but can only export the blocks created by others as json.

File: packages/core-data/src/resolvers.js
Comment (tofumatt) [reply packages/core-data/src/resolvers.js]: You'd need template strings for the first part:

```suggestion
	const path = `/wp/v2/${ resource }` + ( id ? `/${ id }` : '' );
```

I think it's a bit easier to sort out though, yeah.

File: packages/core-data/src/resolvers.js
Comment (tofumatt) [reply packages/core-data/src/resolvers.js]: At the very least could we add the ticket URL in the comment so it can be tracked/checked on?

File: packages/core-data/src/resolvers.js
Comment (tofumatt) [inline packages/core-data/src/resolvers.js]: 😟 This seems like it should throw. Why are we catching an ignoring the error without doing anything with it?

File: packages/core-data/src/test/selectors.js
Comment (tofumatt) [reply packages/core-data/src/test/selectors.js]: Yeah, `true` as the default seems wrong to me. I'd rather UI update after API requests and have extra settings added than them removed.

File: packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js
Comment (tofumatt) [inline packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js]: Any reason to use `!!` here? I just always find it weird to see 🤷‍♂️

File: packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js
Comment (tofumatt) [inline packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js]: Again, any reason for the `!!`?

**Review by tofumatt — DISMISSED** [review]
This is looking good—I like the `canUser` selector API, but I think there's some code cleanup here to do and I have two concerns:

1. why is `true` the default response for `canUser()`?
2. Why are we ignoring the API request throwing an error?

File: packages/core-data/src/resolvers.js
Comment (noisysocks) [reply packages/core-data/src/resolvers.js]: I briefly debugged this when I noticed the issue and from memory it was something to do with `rest_handle_options_request` not calling `$request->set_url_params()` unlike `dispatch` which correctly parses and sets URL parameters.

> Can you provide steps to reproduce?

1. Create a reusable block
1. Perform a `GET http://local.wordpress.test/wp-json/wp/v2/blocks` request and note the ID of the reusable block
1. Perform a `OPTIONS http://local.wordpress.test/wp-json/wp/v2/blocks/27` request, replacing `27` with the ID from step 2. Note that an `Allow` header is not sent back
1. Perform a `GET http://local.wordpress.test/wp-json/wp/v2/blocks/27` request, replacing `27` with the ID from step 2. Note that an `Allow` header is sent back

Comment (noisysocks) [reply packages/core-data/src/resolvers.js]: I created a ticket here: https://core.trac.wordpress.org/ticket/45753

Comment (noisysocks) [reply packages/core-data/src/resolvers.js]: Added a link to the Trac ticket in cf2d7012f.

File: packages/core-data/src/resolvers.js
Comment (noisysocks) [reply packages/core-data/src/resolvers.js]: I noted this in [the agenda for the next meeting](url
https://docs.google.com/document/d/1QADtwgPif4-fCKIlS6rPP9CXhGgWZI2wFx8UfcXf7kk/edit).

File: docs/designers-developers/developers/data/data-core.md
Comment (aduth) [reply docs/designers-developers/developers/data/data-core.md]: > This indentation is kinda odd; I'm not sure why this is indented where it is...

Likely because it's unprocessed from the space-aligned JSDoc. Could probably be improved with some additional processing (`lines.map( trim ).join( ' ' )`?). It has no difference in the Markdown preview though, since the whitespace is collapsed.

File: packages/block-library/src/block/edit.js
Comment (aduth) [reply packages/block-library/src/block/edit.js]: > Any reason you're checking `!! reusableBlock` rather than just using `reusableBlock`? Just curious!

If a value is falsey in this scenario, the specific value is returned in the left-hand of the condition.

If `reusableBlock` was `undefined` for example, without the `!!`, `canUpdateBlock` would be passed as `undefined`, despite its name indicating that it would be a boolean. As written, it satisfies an implicit contract of providing a boolean value. Practically speaking this would rarely have much an impact, except in a scenario where someone expected it to be the particular boolean form (`=== false`).

File: packages/core-data/src/resolvers.js
Comment (noisysocks) [reply packages/core-data/src/resolvers.js]: I like that with the more verbose version that you can glance at the code and see the "shape" of the URL.

File: packages/core-data/src/resolvers.js
Comment (noisysocks) [reply packages/core-data/src/resolvers.js]: Was this suggestion meant to go in this thread? https://github.com/WordPress/gutenberg/pull/12378#discussion_r241679744

File: packages/core-data/src/selectors.js
Comment (noisysocks) [reply packages/core-data/src/selectors.js]: See this thread: https://github.com/WordPress/gutenberg/pull/12378#discussion_r239336329. I think it's fine to leave for now and come back when we have a policy on how to deprecate methods post 5.0.

File: packages/editor/src/components/block-settings-menu/reusable-block-delete-button.js
Comment (noisysocks) [reply packages/editor/src/components/block-settings-menu/reusable-block-delete-button.js]: I like that in this case it lets us avoid passing unnecessary props to the component.

File: packages/core-data/src/resolvers.js
Comment (noisysocks) [reply packages/core-data/src/resolvers.js]: 👍 ccf4936cc

File: packages/core-data/src/resolvers.js
Comment (noisysocks) [reply packages/core-data/src/resolvers.js]: > Should it be named receiveUserPermission using singular version to reflect that you can provide only one permission at the time?

👍 02d9502cd

> At the very least could we add the ticket URL in the comment so it can be tracked/checked on?

I think this is referring to this thread? https://github.com/WordPress/gutenberg/pull/12378#discussion_r239331728 If so I added a link to the Trac ticket in 
https://github.com/WordPress/gutenberg/commit/cf2d7012f039500091aae18ce23e5698f77fa2e0.

File: packages/core-data/src/selectors.js
Comment (noisysocks) [reply packages/core-data/src/selectors.js]: I added a `@deprecated` section to the JSDoc block as part of 62bcb1903606588ca2586bba5cf981843b459261.

File: packages/core-data/src/test/selectors.js
Comment (noisysocks) [reply packages/core-data/src/test/selectors.js]: I've changed `canUser()` to return `false` by default in 62bcb1903606588ca2586bba5cf981843b459261.

File: packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js
Comment (noisysocks) [reply packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js]: 👍 1c3f722b5

File: packages/core-data/src/selectors.js
Comment (noisysocks) [reply packages/core-data/src/selectors.js]: 👍 Done as part of 62bcb1903606588ca2586bba5cf981843b459261.

File: packages/core-data/src/resolvers.js
Comment (noisysocks) [reply packages/core-data/src/resolvers.js]: `error` here refers to the API error that resulted from our OPTIONS request. Doing nothing is correct in this case as we want to leave the store as is. I've added a comment to explain this in e0a7269a1518e89efef5fbd294af9c7417e35c49.

File: packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js
Comment (noisysocks) [reply packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js]: It's to ensure that `typeof isReusable === 'boolean'`.

File: packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js
Comment (noisysocks) [reply packages/editor/src/components/block-settings-menu/reusable-block-convert-button.js]: It's to ensure that `typeof isVisible === 'boolean'`.

**Comment by noisysocks** [comment]
Thanks all for the feedback! I think that I've addressed all outstanding comments.

---

> @noisysocks - nice work on the PR. I personally think it is quite close to be considered ready. My comments are mostly nitpicks. The only important thing from my perspective is to ensure that canUser has the best strategy for returning the default value. It feels like assuming there is no permission is safer.

In 62bcb19 I changed `canUser()` to return `false` by default and added an optional `defaultIsAllowed` param which lets the caller decide what the default return value should be.

> Does this change the UI at all? If so, can you include screenshots and/or GIFs in the description?

Yes. I've added some screenshots to the description.

> At first glance, I'm not sure how well this canUser abstraction is going to work in practice. WordPress' roles and capabilities system is quite nuanced, and abstractions on top of it can be problematic. Just want to flag that.

Could you elaborate on this?

There might be some confusion here. `canUser()` checks a user's ability to perform a particular REST API action on a particular REST API resource. For example, `canUser( 'create' 'media' )` performs an `OPTIONS /wp/v2/media` requests and checks that `POST` appears in the resultant `Allow` header. It **does not** check that a user or a role has a particular capability. 

> 1. why is true the default response for canUser()?
> 2. Why are we ignoring the API request throwing an error?

I've changed the default response of `canUser()` in 62bcb19 and have explained why we ignore the API error in https://github.com/WordPress/gutenberg/pull/12378#discussion_r244897815.

**Comment by noisysocks** [comment]
I think this is ready. Is anyone able to give it a last look over?

File: packages/core-data/src/resolvers.js
Comment (talldan) [reply packages/core-data/src/resolvers.js]: @noisysocks I noticed you `@deprecated` the selector, but the resolver hasn't been deprecated. I guess technically they're the same interface and the resolver wouldn't be called directly, but I wanted to double-check if it was intentional.

Comment (noisysocks) [reply packages/core-data/src/resolvers.js]: That's a good point. I think selectors are the public interface here but it wouldn't hurt to have the `@deprecated` attribute in the resolver as well.

Comment (noisysocks) [reply packages/core-data/src/resolvers.js]: Done in 4ea46a5d6c211e9614e9d2304cf1563a73394fc3!

**Comment by noisysocks** [comment]
Pinging @WordPress/gutenberg-core for a review. Would like to get this in. The current experience in WordPress 5.0. is pretty horrid 🙂

**Comment by noisysocks** [comment]
In 91e49659aba3c0ee947e65dad596579df5c81263 I've used `deprecated()` to mark `hasUploadPermission()` as deprecated as per [the discussion in Slack](https://wordpress.slack.com/archives/C5UNMSU4R/p1547023091120300).

File: packages/editor/src/components/media-placeholder/index.js
Comment (talldan) [inline packages/editor/src/components/media-placeholder/index.js]: This is a personal preference, but I'm not such a big fan of having to pass undefined as an argument here. Some alternatives (but don't feel like it's something that needs to be changed):
```
defaultTo( canUser( 'create', 'media' ), true );
```

```
canUser( [ 'create', 'media' ], true );
```

```
canUser( { action: 'create', resource: 'media', default: true } );
```

**Review by talldan — APPROVED** [review]
Re-tested both media and reusable blocks as a contributor. Works well. 

A++++ great PR

**Comment by talldan** [comment]
IIRC this one also needed a change in core to preload that request.

**Comment by noisysocks** [comment]
Getting this in early so that there's time for it to bake in 5.0.

> IIRC this one also needed a change in core to preload that request.

Thanks for the reminder. I wonder what the best way to do this is... Do we check how the plugin PHP files differ when updating Core's script dependencies? (e.g. `git diff v4.9 v5.0 **.php`) If so, we'll catch this change. If not, perhaps opening a ticket is the safest bet.

**Comment by talldan** [comment]
@noisysocks Not sure, are the files easily diffable, they look quite different. This one was done as a trac ticket, so maybe that's the best option:
https://core.trac.wordpress.org/changeset/43833

File: packages/core-data/package.json
Comment (aduth) [inline packages/core-data/package.json]: There should have been a corresponding change to the root `package-lock.json`. I'm not sure how this passed the build task which checks for local uncommitted changes, but it shouldn't have.

Comment (noisysocks) [reply packages/core-data/package.json]: Huh! Looks like `check-local-changes` doesn't pick this up because `precheck-local-changes` doesn't run `npm install`.

Will fix this up in a seperate PR.

Comment (noisysocks) [reply packages/core-data/package.json]: https://github.com/WordPress/gutenberg/pull/13462

**Comment by paaljoachim** [comment]
A test that I made:
Contributor made a new post. Submitted for Review.
Logged in as admin. Went to the newly created contributor post and took it over. Made a reusable block. Clicked to save as pending.
Logged in as contributor. Went to post. Noticed that contributor could delete the reusable block.

**Comment by talldan** [comment]
@paaljoachim What version are you testing against? This has only just been released in the plugin today afaik.

I couldn't reproduce testing against master.

**Comment by paaljoachim** [comment]
Hey Daniel @talldan

Lets just call it an error on my end. I made the following comment:
https://make.wordpress.org/test/2019/02/05/call-for-testing-gutenberg-5-0/#comment-1170

**Pull request merged**
